src/robot/variables/tablesetter.py

```diff
@@ -45,23 +47,34 @@ def read(self, variables):
                 var.report_invalid_syntax(err)
 
     def _get_name_and_value(self, name, value, error_reporter):
-        return name[2:-1], VariableTableValue(value, name, error_reporter)
+        name, type_ = split_variable_name_with_bracket_and_type(name)
+        return name[2:-1], VariableTableValue(value, name, type_, error_reporter)
 
 
-def VariableTableValue(value, name, error_reporter=None):
+def VariableTableValue(value, name, type_=None, error_reporter=None):
     validate_var(name)
     VariableTableValue = {'$': ScalarVariableTableValue,
                           '@': ListVariableTableValue,
                           '&': DictVariableTableValue}[name[0]]
-    return VariableTableValue(value, error_reporter)
+    return VariableTableValue(value, type_, error_reporter)
 
 
 class VariableTableValueBase(object):
 
-    def __init__(self, values, error_reporter=None):
+    def __init__(self, values, type_=None, error_reporter=None):
         self._values = self._format_values(values)
         self._error_reporter = error_reporter
         self._resolving = False
+        self._type = type_
+
+    def _convert(self, value):
+        try:
+            return convert_variable_value('', value, self._type)
+        except ValueError:
+            pass
+        except TypeError:
+            pass
+        return value
```

Comment:
This doesn't look good. Why are errors silenced? If the idea is to avoid backwards incompatibility problems with cases like `${name:with:colons}`, it would be better to already earlier validate is the part of the last colon a type or not. Blindly ignoring all these exceptions means that bugs in the code are easily hidden as well. Also, an API that requires passing `''` isn't that convenient.

Reply:
I split the PR after you request, this part of code is not anymore